Fix RefCounted releasing early and not clearing reference

rune-scape opened the pull request:
+organized the code
+Ref now releases __After__ aquiring other reference
+fixed case where `reference` returns false should clear the reference

kinda maybe a major bug thats been here since the first commit
probly related to alot of crashes when closing the editor
fixes many cases of `~GDScript()` crashing when cleaning up

(ive been using this code for months now)

rune-scape committed:
Refactor ref-counting code and fix ref counted releasing before aquiring

rune-scape commented:
ive separated one of the fixes into its own commit, but making references release after acquiring is tied into the refactor

ive also refactored the referencing code in variant.cpp into a set of functions that mirror the ones in ref_counted.h

RandomShaper commented:
Thanks. I will need more time to review this properly, but so far I can acknowledge this fixes the race condition were an object being released by another thread is assigned to a `Ref`, making it dangling.

rune-scape commented:
i didnt realize there was another set of functions manipulating refcounts,, i refactored/modified `VariantInternal::object_assign` and other related functions to use the same ones as variant

theres still refcounting code in `CallableCustom` that releases before aquiring, not sure if that should be in this PR ..

RandomShaper reviewed (changes requested):
I've finally been able to give a new review to this. Now I've parsed it more carefully, I see why this is an improvement at several levels.

A few extra requests:
1. I found it a bit hard to tell `ref_pointer_no_init()` from `ref_pointer()`. I'd suggest merging both like this (not tested, but you get the idea; not sure how doable in `Variant::ObjData::ref()`-`ref_pointer()`):
```C++
template<bool Init = False>
_FORCE_INLINE_ void ref_pointer(T *p_refcounted) { // Note 'p_ref` -> 'p_refcounted`
	if (unlikely(p_ref == reference)) {
		return;
	}
	
	// Suggesting this instead of the cleanup_ref approach. Seems more straighforward and avoids dealing with the refcount more than strictly needed. Am I missing something?
	if (reference) {
		reference->unreference();
	}

	bool valid = false;
	if (likely(p_refcounted)) {
		bool valid = false;
		if constexpr (Init) {
			if (p_refcounted->init_ref()) {
				valid = true;
			}
		} else {
			if (p_refcounted->reference()) {
				valid = true;
			}
		}
	}
	reference = likely(valid) ? p_refcounted : nullptr;
}
```
2. As in the code above, it'd be good to use `likely()`/`unlikely()` in the null checks and identity checks throughout. The former code failed to do so, but I think it's worth while you're at it.
3. Maybe opinionated, but I find `ptr` easier to parse than `ptr != null`, especially in this sort of code.
4. In the various functions that are implemented as `*this = p_from;`, I'd suggest being explicit about the call to `ref*()`. After all, anyone trying to wrap their head around this code will have to check how `operator=` is implemented anyway.
5. Now I don't see much reason any longer to have separate commits, unless the first one could be a pure fix and the second one a pure refactor. Otherwise, I'd ask you to squash them back.

I'm aware this looks like a lot of changes, but it's just a little refinement over your work. If you already got close to perfection, these ideas should get us even a step closer!

rune-scape commented:
1. i like this, will make this change, but `Variant::ObjData::ref()` and `ref_pointer()` are too specialized
2. imo this is definitely an over optimization, null references are common and not an error, enforcing cpu branch prediction on common cases can make performance worse in some cases for little to no benefit
3. i think ive been told `ptr == nullptr` is preferred and its what i prefer
4. true
5. ya, i don't think the separate commits makes sense

RandomShaper commented:
1. Any comments on my suggestion about using `reference->unreference()` instead of the cleanup ref approach?
2. OK. - (We'd actually have to judge on each condition separately. Maybe some of the identity checks would benefit? Well, since I can't back this with data, I agree that'd be premature optimization and it's also out of the scope of this PR.)
3. OK. - (My suggestion was more about the not-null checks. I find it harder to parse `!= nullptr` as checking for a _positive_ fact. However, yeah, there's no much point in discussing about that. It's your call.)

akien-mga commented:
> i think ive been told `ptr == nullptr` is preferred and its what i prefer

It's not necessarily consistent throughout the codebase, but I do believe the "preferred" usage currently is `if (ptr)` and `if (!ptr)`. We even used to have some scripts that would remove ` != nullptr` automatically from the codebase, though I think we stopped doing that as it was a bit of a pain and overly pedantic.

I believe what we indicated was that for `Ref<T>` types, we prefer `using the `is_null()` and `is_valid()` methods (which was now made consistent with #96532 to enforce it). But for actual raw pointer types using them as `bool` is the most widespread convention among contributors.

rune-scape commented:
> 1. Any comments on my suggestion about using `reference->unreference()` instead of the cleanup ref approach?

manually calling it doesn't deal with refcount any more than the `cleanup_ref` way, the `cleanup_ref` is default constructed and the data copied
its then destructed (`unref`ed) at the end of the scope, and by using the object lifetime to cleanup i was trying to convey that it should be the last thing done in the function (also the way youve written it would release the reference first, and id call `unref()` instead of `reference->unreference()`)

switched to using `if (ptr)`
rebased to master
i also made the change to Callable to release after aquiring, bc it's very related and could have similar effects

RandomShaper reviewed (approved):
All I can say now is that this is a piece of fine art.

Thank you very much for your patience and diligence dealing with my feedback this long.

akien-mga commented:
Thanks!
